### PR TITLE
enable coverage test for dataloader on Windows

### LINKED ISSUE #50661: test_custom_batch_pin_worker fails on Windows with coverage
## 🐛 Bug

Met when testing https://github.com/pytorch/pytorch/pull/50660. It seems that the test cannot pass when running in coverage mode.
Main process
```pytb
Traceback (most recent call last):
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 422, in wrapper
    fn(*args, **kwargs)
  File "test_dataloader.py", line 2160, in test_custom_batch_pin_worker
    for sample in loader:
  File "C:\Users\circleci\project\build\win_tmp\build\torch\utils\data\dataloader.py", line 517, in __next__
    data = self._next_data()
  File "C:\Users\circleci\project\build\win_tmp\build\torch\utils\data\dataloader.py", line 1182, in _next_data
    idx, data = self._get_data()
  File "C:\Users\circleci\project\build\win_tmp\build\torch\utils\data\dataloader.py", line 1143, in _get_data
    raise RuntimeError('Pin memory thread exited unexpectedly')
RuntimeError: Pin memory thread exited unexpectedly
```
Child process:
```pytb
Traceback (most recent call last):
  File "c:\jenkins\miniconda3\lib\threading.py", line 916, in _bootstrap_inner
    self.run()
  File "c:\jenkins\miniconda3\lib\threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\utils\data\_utils\pin_memory.py", line 25, in _pin_memory_loop
    r = in_queue.get(timeout=MP_STATUS_CHECK_INTERVAL)
  File "c:\jenkins\miniconda3\lib\multiprocessing\queues.py", line 113, in get
    return _ForkingPickler.loads(res)
AttributeError: Can't get attribute 'SimpleCustomBatch' on <module '__main__' from 'C:\\Jenkins\\Miniconda3\\Scripts\\coverage.exe\\__main__.py'>
```
See https://app.circleci.com/pipelines/github/pytorch/pytorch/261028/workflows/ce92c13a-17bb-484f-bd2e-9ff5aa03840c/jobs/10212393

cc @peterjc123 @maxluk @nbcsm @guyang3532 @gunandrose4u @mszhanyi @skyline75489

### PR DESCRIPTION
**Pull request by mszhanyi**
Fixes #50661
For coverage, 
The class qualified name is `'SimpleCustomBatch': <class '__mp_main__.SimpleCustomBatch'>`

For pytest
The class qualified name is `'SimpleCustomBatch': <class 'test_dataloader.SimpleCustomBatch'>`

So move the class to one separate file

image

As @malfet suggestion, use __import__ to avoid adding new file.

### CODE REVIEW

**Comment by mszhanyi** [comment]
https://app.circleci.com/pipelines/github/pytorch/pytorch/275983/workflows/4e70bbf4-d755-4969-8f59-55a98944b66c/jobs/11043974
passed

File: test/run_test.py [hunk 1]
```diff
@@ -340,8 +336,8 @@ def print_to_stderr(message):
     print(message, file=sys.stderr)
 
 
-def get_executable_command(options, allow_pytest, disable_coverage=False):
-    if options.coverage and not disable_coverage:
```
Comment (malfet) [inline test/run_test.py:344]: Hmm, can we please keep the option in place, but simply no longer use it in the run_test.py?

**Review by malfet — CHANGES_REQUESTED** [review]
Thank you for the PR. 
Instead of moving files around, perhaps it would be better to rely on python instrumentation capabilities and dynamically register class in a desired module?
Something along the lines of:
https://github.com/pytorch/pytorch/blob/8af648354f4ae8e00576030f558ee08deb1c4acd/torch/testing/_internal/jit_utils.py#L710-L712

File: test/run_test.py (hunk 1 above)
Comment (mszhanyi) [reply test/run_test.py]: Updated

**Comment by mszhanyi** [comment]
> Thank you for the PR.
> Instead of moving files around, perhaps it would be better to rely on python instrumentation capabilities and dynamically register class in a desired module?
> Something along the lines of:
> https://github.com/pytorch/pytorch/blob/8af648354f4ae8e00576030f558ee08deb1c4acd/torch/testing/_internal/jit_utils.py#L710-L712

Sorry, I don't understand the trick and I tried using setattr but not work.
@malfet , could you please give me an example on the test.

**Comment by malfet** [comment]
I admit this is quite ugly, but works:
```
# Needed to workaround limitations when  `__main__` is not importable my spawned module
self_module = __import__(os.path.splitext(os.path.basename(__file__))[0])


def collate_wrapper(batch):
      return self_module.SimpleCustomBatch(batch)
```

File: test/test_dataloader.py [hunk 2]
```diff
@@ -2064,10 +2063,11 @@ def pin_memory(self):
     def is_pinned(self):
         return self.inp.is_pinned() and self.tgt.is_pinned()
 
-
+# foo is used to fix https://github.com/pytorch/pytorch/issues/50661
+# Needed to workaround limitations when  `__main__` is not importable my spawned module
+foo = __import__(os.path.splitext(os.path.basename(__file__))[0])
```
Comment (malfet) [inline test/test_dataloader.py:2068]:
```suggestion
# Workaround for https://github.com/pytorch/pytorch/issues/50661
# Classes from  `__main__` can not be correctly unpickled from spawned module
# See https://docs.python.org/3/library/multiprocessing.html#multiprocessing-programming
self_module = __import__(os.path.splitext(os.path.basename(__file__))[0])

```

File: test/test_dataloader.py [hunk 3]
```diff
@@ -2064,10 +2063,11 @@ def pin_memory(self):
...
 def collate_wrapper(batch):
-    return SimpleCustomBatch(batch)
-
+    return foo.SimpleCustomBatch(batch)
```
Comment (malfet) [inline test/test_dataloader.py:2070]:
```suggestion
    return self_module.SimpleCustomBatch(batch)

```

File: test/test_dataloader.py [hunk 4]
```diff
@@ -2097,7 +2097,7 @@ def setUp(self):
     @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
     def test_custom_batch_pin(self):
         test_cases = [
-            (collate_wrapper, SimpleCustomBatch),
+            (collate_wrapper, foo.SimpleCustomBatch),
```
Comment (malfet) [inline test/test_dataloader.py:2100]:
```suggestion
            (collate_wrapper, self_module.SimpleCustomBatch),
```

File: test/test_dataloader.py [hunk 5]
```diff
@@ -2111,7 +2111,7 @@ def test_custom_batch_pin(self):
     @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
     def test_custom_batch_pin_worker(self):
         test_cases = [
-            (collate_wrapper, SimpleCustomBatch),
+            (collate_wrapper, foo.SimpleCustomBatch),
```
Comment (malfet) [inline test/test_dataloader.py:2114]:
```suggestion
            (collate_wrapper, self_module.SimpleCustomBatch),
```

**Review by malfet — APPROVED** [review]

**Review by malfet — APPROVED** [review]

**Review by walterddr — APPROVED** [review]

**Pull request closed without merging**